master/internal/hpimportance/manager.go

```diff
@@ -0,0 +1,268 @@
+	"github.com/determined-ai/determined/master/pkg/model"
+)
+
+const (
+	// RootAddr is the path to use for looking up the manager actor.
+	RootAddr = "hpimportance"
+
+	// Evaluate after every 10%, but no more than every 10 minutes
+	minPause   = 10 * time.Minute
+	minPercent = 0.1
+)
+
+// Messages handled by the HP importance manager.
+type (
+	// ExperimentCreated is the message an experiment sends when created.
+	ExperimentCreated struct {
+		ID int
+	}
+
+	// ExperimentCompleted is the message an experiment sends upon completion.
+	ExperimentCompleted struct {
+		ID int
+	}
+
+	// ExperimentProgress is the message an experiment sends after trial completion.
+	ExperimentProgress struct {
+		ID       int
+		Progress float64
+	}
+
+	// WorkRequest is an explicit request to compute HP importance on-demand.
+	WorkRequest struct {
+		ExperimentID int
+		MetricName   string
+		MetricType   model.MetricType
+	}
+)
+
+// The zero-values of these types happen to be sensible defaults too. If we add fields for which
+// that is not true, add a newStateRecord(). This is for state that doesn't need to be persisted,
+// especially if it is access very frequently.
+type stateRecord struct {
+	lastResult   time.Time
+	lastProgress float64
+}
+
+type manager struct {
+	db    *db.PgDB
+	state map[int]stateRecord
+}
+
+// NewManager initializes the master actor (of which there should only be one instance running).
+func NewManager(db *db.PgDB) actor.Actor {
+	return &manager{db, make(map[int]stateRecord)}
+}
+
+func (m *manager) Receive(ctx *actor.Context) error {
+	switch msg := ctx.Message().(type) {
+	case actor.PreStart:
+		// TODO: fetch any pending or in_progress tasks from the DB and trigger them
+	case actor.PostStop:
+		// Do nothing
+	case actor.ChildFailed:
+		ctx.Log().Warnf("hyperparameter importance worker failed: %+v", msg)
+	case actor.ChildStopped:
+		// Do nothing - it'll respawn next time a request is received
+	case ExperimentCompleted:
+		m.experimentCompleted(ctx, msg)
+	case ExperimentCreated:
+		m.experimentCreated(ctx, msg)
+	case ExperimentProgress:
+		m.experimentProgress(ctx, msg)
+	case WorkRequest:
+		m.workRequest(ctx, msg)
+	case workStarted:
+		m.workStarted(ctx, msg)
+	case workFailed:
+		m.workFailed(ctx, msg)
+	case workCompleted:
+		m.workCompleted(ctx, msg)
+	default:
+		ctx.Log().Errorf("unknown message received by hyperparameter importance manager: %v!",
+			ctx.Message())
+	}
+	return nil
+}
+
+func (m *manager) experimentCompleted(ctx *actor.Context, msg ExperimentCompleted) {
+	m.triggerDefaultWork(ctx, msg.ID)
+}
+
+func (m *manager) experimentCreated(ctx *actor.Context, msg ExperimentCreated) {
+	m.state[msg.ID] = stateRecord{
+		lastResult: time.Now(),
+	}
+}
+
+func (m *manager) experimentProgress(ctx *actor.Context, msg ExperimentProgress) {
+	state, ok := m.state[msg.ID]
+	if !ok {
+		state = stateRecord{}
+	}
+	if msg.Progress-state.lastProgress > minPercent &&
```

Comment:
Yeah we discussed this in one of our meetings (I *think* it was with Liam)? I believe this is neither critical nor fragile. We're just looking for 10% milestones. It's not always monotonic, or goes out of order, it is of little consequence. There's a 10 minute minimum to make sure it's not happening constantly (and experiments that would take minutes to run this on would hit 10% milestones after days, not minutes) because of fluctuations or jumps. And this is also purely for convenience: especially now that we've already added the ability to request results on-demand, it's not critical.

Reply:
Ok, that sounds good to me. I just didn't stop to give it a ton of thought earlier and wanted to make sure we did that at some point.